### ISSUE TITLE
Error "opcode #38 (report please)" when to a custom class

### ISSUE THREAD

**Issue by theraot** [godotengine/godot#43042]
I found another report of opcode 38: #43023

**Godot version:**
3.2.4.beta1

**OS/device including version:**
Windows 10

**Issue description:**
I created a scene with root node with a script attached that has a custom class. I instantiated the scene, and tried to cast to the custom class. Expected the cast to succeed. Got the following error:

```
Internal Script Error! - opcode #38 (report please).
```

Errors panel says:

```
E 0:00:00.757   call: Condition ' !base_type ' is true. Breaking..:
  <Fuente C++>  modules/gdscript/gdscript_function.cpp:938 @ call()
  <Stack Trace> Core.gd:6 @ _ready()
```

**Steps to reproduce:**

1. Create a new project. Add two (2D) scenes. Let us call them "Main.tscn" and "Item.tscn"
2. In Item.tscn, attach a script to the root note (A Node2D), with this code:

```
extends Node2D

class_name Item
```

3. In Main.tscn, add a node with an attached script, with this code:

```
extends Node2D

onready var item = load("Item.tscn");

func _ready():
    var created_item = item.instance() as Item; #error here
    add_child(created_item);
```

This can be worked around by not casting. Without casting, we can still use `call` et.al. to interact with the script.

**Minimal reproduction project:**

[Test.zip](https://github.com/godotengine/godot/files/5431493/Test.zip)

**Comment by theraot** [comment]
@Calinou It is worth noting that the error on #34023 says:

```
modules/gdscript/gdscript_function.cpp:837 @ call()
```

While this one says:

```
modules/gdscript/gdscript_function.cpp:938 @ call()
```

Line `837` vs line `938`.

**Comment by theraot** [comment]
After having a look at the project from #43023, I tried and found that when I change the code in my project to the following you get the error on line 387:

```
extends Node2D

onready var item = load("Item.tscn");

func _ready():
    var tmp = item.instance();
    var created_item: Item = tmp; #error here
    add_child(created_item);
```

**Comment by Calinou** [comment]
@theraot If you can build Godot from source, can you try to [bisect the regression](https://docs.godotengine.org/en/latest/community/contributing/bisecting_regressions.html)? This would be hugely appreciated :slightly_smiling_face:

**Comment by theraot** [comment]
@Calinou I'll try. I have not been able to built Godot from source yet. In fact, After my last post here I decided to follow [Compiling for Windows](https://docs.godotengine.org/en/latest/development/compiling/compiling_for_windows.html). I think my scons setup failed, haven't figured out why yet.

**Comment by theraot** [comment]
I built commit a03af876d1a8c30723021bc60ce10d38d2a8d534 (current head of 3.2 branch) 

---

This code

```
extends Node2D

onready var item = load("Item.tscn");

func _ready():
    var created_item = item.instance() as Item; #error here
    add_child(created_item);
```

Gets:

```
Internal Script Error! - opcode #17 (report please).
```

```
E 0:00:00.644   GDScriptFunction::call: Condition ' !base_type ' is true. Breaking..:
  <Fuente C++>  modules\gdscript\gdscript_function.cpp:938 @ GDScriptFunction::call()
  <Stack Trace> Core.gd:6 @ _ready()
```
---

This code:

```
extends Node2D

onready var item = load("Item.tscn");

func _ready():
    var tmp = item.instance();
    var created_item: Item = tmp; #error here
    add_child(created_item);
```

Gets:

```
Internal Script Error! - opcode #14 (report please).
```

```
E 0:00:00.599   GDScriptFunction::call: Condition ' !base_type ' is true. Breaking..:
  <Fuente C++>  modules\gdscript\gdscript_function.cpp:837 @ GDScriptFunction::call()
  <Stack Trace> Core.gd:7 @ _ready()
```
---

I'll try to find out the last commit without this issue.

**Comment by Error7Studios** [comment]
Found that this works with `Node`, `Control`, and `Node2D` classes.
As stated above, using `static typing` causes it at line `837`, and  `casting` causes the bug at .cpp line `938`.

![opcode38](https://user-images.githubusercontent.com/60025044/97066505-c0847900-157b-11eb-8c23-65a56b052f41.png)

**Comment by theraot** [comment]
79f15027fc12685c931b3c2383bc7aaf771dc7ba Works
0d8b2d34e1637cca7f7d3efab69c1e7a69705f03, 6dffc1ef46bcfba4e904a5e5850a3fffb5e3b54c Fails

I'm getting close.

**Comment by Error7Studios** [comment]
Don't know if it helps or not, but I just found that `load()` triggers the error, but `preload()` doesn't

**Comment by theraot** [comment]
@Error7Studios Perhaps.

My understanding of Godot source is very limited. However, my current hypothesis is that the class is not being registered correctly, and thus when the runtime tries to check the type for the cast, it fails. Perhaps `preloads` does it correctly and `load` does not.

Once I figure out what commit introduced the problem, I'd have a better idea of what it is.

**Comment by theraot** [comment]
f388ea00bf820d8bfa621a9ff394804751db91e9 Works
76a84587535f478551e9bdf2c11d6a710a7a322b, aa0221f0fba194bf996bbcde290342f969a5319c Fails

**Comment by theraot** [comment]
I can confirm that bca2633f93fcb2f44c176c6d513e666ab8153c2b introduced the problem.

I can also confirm that using `preload` avoids the problem.

bca2633f93fcb2f44c176c6d513e666ab8153c2b is intended to fix a leak caused by cyclic references. I guess it did not touch `preload`, it didn't have to, since `preload` won't work when there are cyclic references. That explains that only `load` is affected.

Looking at the diff, I guess the `Ref<Script>` is causing the checks at gdscript_function.cpp to trip.

---

About fixing it, I don't know, but I have the source and the means to build it, I might give it go.

**Comment by theraot** [comment]
CC @RandomShaper

**Comment by ChainedLupine** [comment]
Chiming in, as the "!base_type is true/ISE opcode #17" error is what got me involved.

**Comment by akien-mga** [comment]
Fixed by #43049.
